spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/actions/BaseDeleteReachableFilesSparkAction.java

```diff
@@ -125,8 +126,9 @@ private Result doExecute() {
 
   private Dataset<Row> buildReachableFileDF(TableMetadata metadata) {
     Table staticTable = newStaticTable(metadata, io);
-    return withFileType(buildValidContentFileDF(staticTable), CONTENT_FILE)
-        .union(withFileType(buildManifestFileDF(staticTable), MANIFEST))
+    Dataset<Row> allManifests = loadMetadataTable(staticTable, ALL_MANIFESTS);
+    return withFileType(buildValidContentFileDF(staticTable, allManifests), CONTENT_FILE)
+        .union(withFileType(buildManifestFileDF(allManifests), MANIFEST))
```

Comment:
How does this change the performance? Don't we have to compute all manifests in both locations here? Or Does changing to an object let Spark cache the relation?

Reply:
https://github.com/apache/iceberg/blob/434c5658b26ff06f11da261d1d49a84f19e267d7/core/src/main/java/org/apache/iceberg/AllManifestsTable.java#L118

For example in expire_snapshots, as the ALL_MANIFESTS metadata tables computed twice, it has to read manifest file list twice for every snapshot. Now only once ALL_MANIFESTS metadata table is created and it is reused.

Reply:
`loadMetadataTable` was called twice, now it is only once for ALL_MANIFESTS.

Reply:
> Or Does changing to an object let Spark cache the relation?

I thought compute of dataset will happen in the first action and the results are reused for both the steps.

Reply:
@RussellSpitzer : I got what you meant now.

To solve this problem only way I have now is to collect the dataset as list of rows and prepare new dataset from list of rows and use that dataset in both location. This way, reading of manifest list happens only once. WDYT ? any other solution?

Reply:
fixed and verified the scanning by adding and checking logs.

Reply:
It seems, individual test cases passes. But when whole suit is ran, persist() is causing failures for hadoop catalog. Not sure why. I might have to explore more on this. Probably I need some job completion listener to unpersist().

Reply:
Ah yes I should have been more clear, I was referring to the fact that the dataset would be recomputed on both lines. The "loadMetadataTable" function should be very fast but the actual planning should be expensive and that would require a cache of some kind. 

I'm a little worried in general about persisting things since I want to make sure we clean up our caches asap.

Reply:
> Very much dependent on the computation leading up to the cached df

Because it involves the IO operation, it will definitely help when there are hundreds or thousands of snapshots.

Reply:
@RussellSpitzer , which is better? a cache or manually calling dataset.collect() on that allManifest df and building new dataset on top of those collected rows and reusing it on both the location?

Reply:
@ajantha-bhat The problem is that persisting does another IO serialization of what should be basically the same amount of information but hopefully in a more readable form. Persist by default is a Memory and Disk based cached. You really need to test it out to be sure.

For Cache vs Collect. Cache is probably much better, pieces would be stored on executors and the IO would hopefully be mostly local, doing a Collect and and building a DF from it would essentially bring all data back to the driver serialize, then deserialize and send everything back out. This is always worse than cache/persist